src/loader.rs

```diff
@@ -11,11 +16,243 @@ impl Loader {
     }
   }
 
-  pub(crate) fn load<'src>(&'src self, path: &Path) -> RunResult<&'src str> {
+  pub(crate) fn load_without_includes<'src>(&'src self, path: &Path) -> RunResult<&'src str> {
     let src = fs::read_to_string(path).map_err(|io_error| Error::Load {
       path: path.to_owned(),
       io_error,
     })?;
     Ok(self.arena.alloc(src))
   }
+
+  pub(crate) fn load_with_includes<'src>(&'src self, path: &Path) -> RunResult<&'src str> {
+    let src = self.perform_load(path, HashSet::new())?;
+    Ok(self.arena.alloc(src))
+  }
+
+  fn perform_load(&self, path: &Path, seen_paths: HashSet<PathBuf>) -> RunResult<String> {
```

Comment:
Not sure about the name, how about:

```suggestion
  fn load_with_includes_recursive(&self, path: &Path, seen_paths: HashSet<PathBuf>) -> RunResult<String> {
```